skynet.go

```diff
@@ -57,33 +132,259 @@ type (
+	}
+
 	// DefaultDownloadOptions contains the default download options.
 	DefaultDownloadOptions = DownloadOptions{
-		PortalURL:          "https://siasky.net",
+		ConnectionOptions:  DefaultConnectionOptions,
 		PortalDownloadPath: "/",
 	}
 
 	// DefaultUploadOptions contains the default upload options.
 	DefaultUploadOptions = UploadOptions{
-		PortalURL:                    "https://siasky.net",
+		ConnectionOptions:            DefaultConnectionOptions,
 		PortalUploadPath:             "/skynet/skyfile",
 		PortalFileFieldName:          "file",
 		PortalDirectoryFileFieldName: "files[]",
 	}
+
+	// ErrResponseError is the error for a response with a status code >= 400.
+	ErrResponseError = errors.New("error response")
 )
 
+// AddSkykey stores the given base-64 encoded skykey with the renter's skykey
+// manager.
+func AddSkykey(skykey string, opts AddSkykeyOptions) error {
+	body := &bytes.Buffer{}
+	url := makeURL(opts.PortalURL, opts.PortalAddSkykeyPath)
+	url = fmt.Sprintf("%s?skykey=%s", url, skykey)
+
+	req, err := makeRequest(opts.ConnectionOptions, "POST", url, body)
+	if err != nil {
+		return errors.AddContext(err, "could not make request")
+	}
+
+	// Add the skykey.
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return errors.AddContext(err, "could not execute POST")
+	}
+	if resp.StatusCode >= 400 {
+		return errors.AddContext(makeResponseError(resp), "error code received")
+	}
+
+	return nil
+}
+
+// CreateSkykey returns a new skykey created and stored under the given name
+// with the given type. skykeyType can be either "public-id" or "private-id".
+func CreateSkykey(name, skykeyType string, opts CreateSkykeyOptions) (Skykey, error) {
+	body := &bytes.Buffer{}
+	url := makeURL(opts.PortalURL, opts.PortalCreateSkykeyPath)
+	url = fmt.Sprintf("%s?name=%s&type=%s", url, name, skykeyType)
+
+	req, err := makeRequest(opts.ConnectionOptions, "POST", url, body)
+	if err != nil {
+		return Skykey{}, errors.AddContext(err, "could not make request")
+	}
+
+	// Create the skykey.
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return Skykey{}, errors.AddContext(err, "could not execute POST")
+	}
+	if resp.StatusCode >= 400 {
+		return Skykey{}, errors.AddContext(makeResponseError(resp), "error code received")
+	}
+
+	// parse the response
+	body = &bytes.Buffer{}
+	_, err = body.ReadFrom(resp.Body)
+	if err != nil {
+		return Skykey{}, errors.AddContext(err, "could not parse response body")
+	}
+	err = resp.Body.Close()
+	if err != nil {
+		return Skykey{}, errors.AddContext(err, "could not close response body")
+	}
+
+	var apiResponse CreateSkykeyResponse
+	err = json.Unmarshal(body.Bytes(), &apiResponse)
+	if err != nil {
+		return Skykey{}, errors.AddContext(err, "could not unmarshal response JSON")
+	}
+
+	return Skykey(apiResponse), nil
+}
+
+// GetSkykey returns the given skykey. One of either name or id must be provided
+// -- the one that is not provided should be "".
```

Comment:
```suggestion
// -- the one that is not provided should be left blank.
```

should be "" made me think i had to set it to "" when I submitted it.